In [23]:
import pandas as pd
import numpy as np
import os
import unicodedata

In [43]:
REPO_URL = 'https://github.com/LuisBetancourt11/Proyecto_Ciencia_Datos'
REPO_DIR = 'Proyecto_Ciencia_Datos'

if not os.path.isdir(REPO_DIR):
    !git clone {REPO_URL}
else:
    print("El repositorio ya está clonado")
    !cd {REPO_DIR} && git pull

El repositorio ya está clonado
Already up to date.


In [25]:
raw_dir = REPO_DIR + '/data/raw/'
processed_dir = REPO_DIR + '/data/processed/'

# Asegurar que la carpeta de salida exista
os.makedirs(processed_dir, exist_ok=True)

print("raw_dir      :", raw_dir)
print("processed_dir:", processed_dir)
print("¿Existe el CSV?:", os.path.exists(raw_dir + 'matriculas.csv'))

raw_dir      : Proyecto_Ciencia_Datos/data/raw/
processed_dir: Proyecto_Ciencia_Datos/data/processed/
¿Existe el CSV?: True


In [26]:
df = pd.read_csv(raw_dir + 'matriculas.csv')
print(df.shape)
df.head()

(13461, 21)


,Género,Municipio de Nacimiento,País,Municipio Dirección,Barrio Dirección,Estrato,Estado civil,EPS,Grupo Sisben,Zona,...,Ocupación,Medio de Transporte,Multiculturalidad,Estado,Fecha de Matrícula,Jornada,Programa,Período,Nivel,Edad
0,Femenino,Medellín,Colombia,Envigado,Las Flores,3.0,Soltero(a),EPS Sura,Sin respuesta,Urbana,...,Estudiante básica,A Pie,No aplica,Graduado,2018-02-05T10:00:00Z,MEDIA TÉCNICA,MEDIA TÉCNICA TECNICO LABORAL AUXILIAR ADMINIS...,2018-2,SEMESTRE 2,20.0
1,Femenino,Envigado,Colombia,Envigado,San José,2.0,Soltero(a),EPS Sura,Sin respuesta,Urbana,...,Estudiante superior,Público Metro,No aplica,Graduado,2016-01-19T10:00:00Z,NOCHE,10T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO...,2016-2,SEMESTRE 2,29.0
2,Masculino,Bogotá,Colombia,La Mesa,Sin respuesta,3.0,Soltero(a),No Aplica,Sin respuesta,Rural,...,Estudiante superior,Transmilenio,No aplica,Cancelado,2021-02-08T10:00:00Z,NOCHE,09T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO,2021-2,SEMESTRE 1,19.0
3,Masculino,Envigado,Colombia,Envigado,San José,2.0,Soltero(a),SISBEN,B3,Urbana,...,Estudiante superior,Público Bus o Buseta,No aplica,Graduado,2020-01-17T10:00:00Z,DIURNA,01T-TÉCNICO LABORAL AUXILIAR DE TALENTO HUMANO,2021-1,SEMESTRE 2,21.0
4,Masculino,Medellín,Colombia,Envigado,El Salado,2.0,Soltero(a),Nueva Promotora de Salud - Nueva EPS,N0,Urbana,...,Estudiante superior,Moto,No aplica,Inactivo,2020-01-27T10:00:00Z,NOCHE,27T-TÉCNICO LABORAL EN MERCADEO Y VENTAS (Estu...,2020-1,SEMESTRE 1,20.0


In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13461 entries, 0 to 13460
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Género                   13461 non-null  object 
 1   Municipio de Nacimiento  13461 non-null  object 
 2   País                     13461 non-null  object 
 3   Municipio Dirección      13407 non-null  object 
 4   Barrio Dirección         13461 non-null  object 
 5   Estrato                  12193 non-null  float64
 6   Estado civil             13461 non-null  object 
 7   EPS                      13460 non-null  object 
 8   Grupo Sisben             13461 non-null  object 
 9   Zona                     12160 non-null  object 
 10  Nivel de formación       13461 non-null  object 
 11  Ocupación                11543 non-null  object 
 12  Medio de Transporte      13461 non-null  object 
 13  Multiculturalidad        13461 non-null  object 
 14  Estado                

## Estandarización de nombres de columnas

Las columnas vienen con tildes, mayúsculas y espacios (`Municipio de Nacimiento`,
`Fecha de Matrícula`, ...). Las normalizamos a `snake_case` sin tildes para trabajar
más cómodamente y evitar errores de tipeo. Se hace de forma programática (no a mano)
porque son 21 columnas.


In [44]:
def normalizar(texto):
    """minúsculas, sin tildes, espacios -> guion bajo"""
    texto = texto.strip().lower()
    # quitar tildes / acentos
    texto = unicodedata.normalize('NFKD', texto).encode('ascii', 'ignore').decode('ascii')
    texto = texto.replace(' ', '_')
    return texto

df.columns = [normalizar(c) for c in df.columns]
df.columns.tolist()

['genero',
 'municipio_de_nacimiento',
 'pais',
 'municipio_direccion',
 'barrio_direccion',
 'estrato',
 'estado_civil',
 'eps',
 'grupo_sisben',
 'zona',
 'nivel_de_formacion',
 'ocupacion',
 'medio_de_transporte',
 'estado',
 'fecha_de_matricula',
 'jornada',
 'programa',
 'periodo',
 'nivel',
 'edad',
 'anio_matricula',
 'mes_matricula',
 'desercion']

## Tratamiento de nulos "disfrazados"

En este dataset los faltantes no vienen como `NaN`, sino como texto:
`Sin respuesta`, `No Aplica`, `No aplica`, etc. Si no los convertimos a nulos reales,
`df.isnull()` no los detecta y contaminan el análisis.

Primero medimos cuántos nulos reales hay **antes** de la conversión.

In [29]:
df.isnull().sum()

,0
genero,0
municipio_de_nacimiento,0
pais,0
municipio_direccion,54
barrio_direccion,0
estrato,1268
estado_civil,0
eps,1
grupo_sisben,0
zona,1301


## Tratamiento de nulos "disfrazados"

En este dataset los faltantes no vienen como `NaN`, sino como texto:
`Sin respuesta`, `No Aplica`, `No aplica`, etc. Si no los convertimos a nulos reales,
`df.isnull()` no los detecta y contaminan el análisis.

In [47]:
df.isnull().sum()

,0
genero,0
municipio_de_nacimiento,1128
pais,51
municipio_direccion,53
barrio_direccion,1318
estrato,1210
estado_civil,1436
eps,3030
grupo_sisben,10285
zona,1269


In [49]:
# Cadenas que representan ausencia de dato
sentinelas = ['sin respuesta', 'sin informacion', 'sin información',
              'no aplica', 'no aplica.', 'na', 'n/a', 'sin dato', 'ninguno']

obj_cols = df.select_dtypes(include='object').columns
for col in obj_cols:
    mask = df[col].astype(str).str.strip().str.lower().isin(sentinelas)
    df.loc[mask, col] = np.nan

print("Nulos reales tras normalizar los textos faltantes:")
df.isnull().sum().sort_values(ascending=False)

Nulos reales tras normalizar los textos faltantes:


,0
grupo_sisben,10285
nivel_de_formacion,3335
medio_de_transporte,3304
eps,3030
ocupacion,1906
estado_civil,1436
barrio_direccion,1318
zona,1269
anio_matricula,1211
fecha_de_matricula,1211


## Estandarización de tipos de datos

- `estrato` y `edad` vienen como `float` (3.0, 20.0). Los pasamos a entero anulable
  (`Int64`), que admite nulos.
- `fecha_de_matricula` viene como texto ISO (`2018-02-05T10:00:00Z`). La parseamos a
  datetime.

In [50]:
for col in ['estrato', 'edad']:
    if col in df.columns:
        df[col] = pd.to_numeric(df[col], errors='coerce').astype('Int64')

df['fecha_de_matricula'] = pd.to_datetime(df['fecha_de_matricula'], errors='coerce',
                                          utc=True).dt.tz_localize(None)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13397 entries, 0 to 13396
Data columns (total 23 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   genero                   13397 non-null  object        
 1   municipio_de_nacimiento  12269 non-null  object        
 2   pais                     13346 non-null  object        
 3   municipio_direccion      13344 non-null  object        
 4   barrio_direccion         12079 non-null  object        
 5   estrato                  12187 non-null  Int64         
 6   estado_civil             11961 non-null  object        
 7   eps                      10367 non-null  object        
 8   grupo_sisben             3112 non-null   object        
 9   zona                     12128 non-null  object        
 10  nivel_de_formacion       10062 non-null  object        
 11  ocupacion                11491 non-null  object        
 12  medio_de_transporte      10093 n

# **Limpieza de datos invalidos y normalizar categorías**

# ***El estrato 0***
En Colombia los estratos socioeconómicos válidos van de **1 a 6**; el valor `0` es un error
de registro (46 casos) y además hay 1,268 registros sin estrato. **Diagnóstico:** esos nulos
no son aleatorios — el 90% de los cursos de extensión no tiene estrato (no se recogió el
formulario socioeconómico), mientras que en técnicos laborales el faltante es solo 1.2%.

Por lo tanto ambos casos se imputan con la **mediana** de los valores válidos, y se crea la
bandera `estrato_imputado` (1 = el valor fue imputado) para que el modelo no pierda la señal
de que el dato no estaba registrado.
Se usa la mediana y no el promedio porque el estrato es una variable ordinal: la mediana
siempre devuelve un estrato que existe

# ***Medio de transporte***

La columna viene con 20 categorías las cuales son duplicadas, por lo tanto se agruparan en 8 variables estandar con lo que reduciriamos el ruido y facilitaría la codificación del modelo.

In [51]:
# --- 1) Estrato: imputar invalidos (0) y faltantes (NaN) con la mediana + bandera ---
# Diagnóstico previo: los NaN de estrato NO son aleatorios — se concentran en cursos de
# extensión (90% sin estrato) donde no se recogió el formulario socioeconómico completo.
# Por eso, además de imputar, se crea la bandera `estrato_imputado`: conserva la señal de
# "dato no registrado", que puede ser predictiva por sí misma.

n_estrato_0  = (df['estrato'] == 0).sum()
n_estrato_na = df['estrato'].isna().sum()

# La mediana se calcula SOLO con valores válidos (1-6): ni ceros ni nulos contaminan el cálculo
mediana_estrato = df.loc[df['estrato'] > 0, 'estrato'].median()

df['estrato_imputado'] = (df['estrato'].isna() | (df['estrato'] == 0)).astype(int)
df.loc[df['estrato'] == 0, 'estrato'] = int(mediana_estrato)
df['estrato'] = df['estrato'].fillna(int(mediana_estrato))

print(f"Mediana usada: {int(mediana_estrato)}")
print(f"Imputados por estrato 0 (inválido): {n_estrato_0}")
print(f"Imputados por NaN (sin registro):   {n_estrato_na}")
print()
print("Distribución de estrato tras la imputación (ya sin nulos):")
print(df['estrato'].value_counts().sort_index())
print()
print("Bandera estrato_imputado:")
print(df['estrato_imputado'].value_counts())

# --- 2) Normalizar medio de transporte ---
mapa_transporte = {
    # A pie
    'A Pie': 'A pie', 'Peatón': 'A pie',
    # Transporte público colectivo
    'Público Bus o Buseta': 'Transporte público', 'SITP o Bus': 'Transporte público',
    'Transmilenio': 'Transporte público', 'Masivo Integrado de Occidente': 'Transporte público',
    'Público Metro': 'Transporte público', 'Público Otro': 'Transporte público',
    # Vehículo propio
    'Moto': 'Moto', 'Particular Motocicleta': 'Moto',
    'Particular Automóvil': 'Automóvil', 'Carro Particular': 'Automóvil',
    'Bicicleta': 'Bicicleta', 'Particular Bicicleta': 'Bicicleta',
    # Taxi
    'Taxi': 'Taxi', 'Público Taxi': 'Taxi',
    # Otros
    'Ruta Escolar': 'Ruta escolar',
    'Otro': 'Otro', 'Particular Otro': 'Otro',
}

Mediana usada: 3
Imputados por estrato 0 (inválido): 45
Imputados por NaN (sin registro):   1210

Distribución de estrato tras la imputación (ya sin nulos):
estrato
1     392
2    5325
3    6843
4     672
5     138
6      27
Name: count, dtype: Int64

Bandera estrato_imputado:
estrato_imputado
0    12142
1     1255
Name: count, dtype: int64


## Columnas derivadas de la fecha

Extraemos componentes de la fecha que pueden ser útiles
para el EDA y el modelado.

In [32]:
df['anio_matricula'] = df['fecha_de_matricula'].dt.year
df['mes_matricula']  = df['fecha_de_matricula'].dt.month
df[['fecha_de_matricula', 'anio_matricula', 'mes_matricula']].head()

,fecha_de_matricula,anio_matricula,mes_matricula
0,2018-02-05 10:00:00,2018.0,2.0
1,2016-01-19 10:00:00,2016.0,1.0
2,2021-02-08 10:00:00,2021.0,2.0
3,2020-01-17 10:00:00,2020.0,1.0
4,2020-01-27 10:00:00,2020.0,1.0


In [33]:
df

,genero,municipio_de_nacimiento,pais,municipio_direccion,barrio_direccion,estrato,estado_civil,eps,grupo_sisben,zona,...,multiculturalidad,estado,fecha_de_matricula,jornada,programa,periodo,nivel,edad,anio_matricula,mes_matricula
0,Femenino,Medellín,Colombia,Envigado,Las Flores,3,Soltero(a),EPS Sura,NaN,Urbana,...,NaN,Graduado,2018-02-05 10:00:00,MEDIA TÉCNICA,MEDIA TÉCNICA TECNICO LABORAL AUXILIAR ADMINIS...,2018-2,SEMESTRE 2,20,2018.0,2.0
1,Femenino,Envigado,Colombia,Envigado,San José,2,Soltero(a),EPS Sura,NaN,Urbana,...,NaN,Graduado,2016-01-19 10:00:00,NOCHE,10T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO...,2016-2,SEMESTRE 2,29,2016.0,1.0
2,Masculino,Bogotá,Colombia,La Mesa,NaN,3,Soltero(a),NaN,NaN,Rural,...,NaN,Cancelado,2021-02-08 10:00:00,NOCHE,09T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO,2021-2,SEMESTRE 1,19,2021.0,2.0
3,Masculino,Envigado,Colombia,Envigado,San José,2,Soltero(a),SISBEN,B3,Urbana,...,NaN,Graduado,2020-01-17 10:00:00,DIURNA,01T-TÉCNICO LABORAL AUXILIAR DE TALENTO HUMANO,2021-1,SEMESTRE 2,21,2020.0,1.0
4,Masculino,Medellín,Colombia,Envigado,El Salado,2,Soltero(a),Nueva Promotora de Salud - Nueva EPS,N0,Urbana,...,NaN,Inactivo,2020-01-27 10:00:00,NOCHE,27T-TÉCNICO LABORAL EN MERCADEO Y VENTAS (Estu...,2020-1,SEMESTRE 1,20,2020.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13456,Masculino,Envigado,Colombia,Envigado,Barrio Mesa,3,Soltero(a),21) EPS Sura,D21,Urbana,...,NaN,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN
13457,Masculino,NaN,Colombia,Envigado,La Magnolia,3,NaN,COOSALUD EPS,B3,Urbana,...,Afrodescendiente,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN
13458,Masculino,NaN,Colombia,Envigado,San José,0,NaN,NaN,NaN,NaN,...,NaN,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN
13459,Femenino,Envigado,Colombia,Envigado,La Mina,3,Soltero(a),21) EPS Sura,N0,Urbana,...,NaN,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN


### Definición adoptada (criterio conservador + regla temporal)
> Se considera **deserción (`1`)** cuando hay **abandono explícito** de los estudios, o
> cuando un estudiante en estado *temporal* (inactivo/aplazado) lleva **más de 2 años** sin
> volver a matricularse
> si tras cierto tiempo no hay reingreso, se considera abandono. Se considera **no deserción
> (`0`)** cuando el estudiante continúa, culminó, o lleva **menos de 2 años** en estado
> temporal (aún podría reingresar).

| Estado original | Clasificación | Justificación |
|---|---|---|
| Graduado, Egresado | **0** | Culminó. *Egresado y Graduado se tratan como equivalentes.* |
| Activo | **0** | Continúa matriculado. |
| Cancelado, Desertor, Retiro - Aplazados, Sancionado | **1** | Abandono explícito / no continuación. |
| **Inactivo, Aplazamiento Semestre, Aplazamiento Actividades** | **regla de 2 años** | `1` si la última matrícula fue hace **más de 2 años**; `0` si fue hace **menos de 2 años**. |
| Sin estado | **excluido (NaN)** | Sin información. |


In [34]:
# Valores reales de la variable de estado ANTES de mapear
df['estado'].value_counts(dropna=False)

,count
estado,
Inactivo,4838
Graduado,3831
Activo,2808
Cancelado,759
Sancionado,602
Aplazamiento Semestre,285
Sin estado,243
Desertor,79
Aplazamiento Actividades,8


In [35]:
mapa_desercion = {
    # --- No deserción (0): continúa o culminó ---
    'Graduado': 0,
    'Egresado': 0,
    'Activo':   0,
    # --- Deserción (1): abandono explícito ---
    'Cancelado':          1,
    'Desertor':           1,
    'Retiro - Aplazados': 1,
    'Sancionado':         1,
    # --- Sin información: se excluye ---
    'Sin estado': np.nan,
}

df['desercion'] = df['estado'].map(mapa_desercion)

In [36]:
df

,genero,municipio_de_nacimiento,pais,municipio_direccion,barrio_direccion,estrato,estado_civil,eps,grupo_sisben,zona,...,estado,fecha_de_matricula,jornada,programa,periodo,nivel,edad,anio_matricula,mes_matricula,desercion
0,Femenino,Medellín,Colombia,Envigado,Las Flores,3,Soltero(a),EPS Sura,NaN,Urbana,...,Graduado,2018-02-05 10:00:00,MEDIA TÉCNICA,MEDIA TÉCNICA TECNICO LABORAL AUXILIAR ADMINIS...,2018-2,SEMESTRE 2,20,2018.0,2.0,0.0
1,Femenino,Envigado,Colombia,Envigado,San José,2,Soltero(a),EPS Sura,NaN,Urbana,...,Graduado,2016-01-19 10:00:00,NOCHE,10T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO...,2016-2,SEMESTRE 2,29,2016.0,1.0,0.0
2,Masculino,Bogotá,Colombia,La Mesa,NaN,3,Soltero(a),NaN,NaN,Rural,...,Cancelado,2021-02-08 10:00:00,NOCHE,09T-TÉCNICO LABORAL EN AUXILIAR ADMINISTRATIVO,2021-2,SEMESTRE 1,19,2021.0,2.0,1.0
3,Masculino,Envigado,Colombia,Envigado,San José,2,Soltero(a),SISBEN,B3,Urbana,...,Graduado,2020-01-17 10:00:00,DIURNA,01T-TÉCNICO LABORAL AUXILIAR DE TALENTO HUMANO,2021-1,SEMESTRE 2,21,2020.0,1.0,0.0
4,Masculino,Medellín,Colombia,Envigado,El Salado,2,Soltero(a),Nueva Promotora de Salud - Nueva EPS,N0,Urbana,...,Inactivo,2020-01-27 10:00:00,NOCHE,27T-TÉCNICO LABORAL EN MERCADEO Y VENTAS (Estu...,2020-1,SEMESTRE 1,20,2020.0,1.0,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13456,Masculino,Envigado,Colombia,Envigado,Barrio Mesa,3,Soltero(a),21) EPS Sura,D21,Urbana,...,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN,0.0
13457,Masculino,NaN,Colombia,Envigado,La Magnolia,3,NaN,COOSALUD EPS,B3,Urbana,...,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN,0.0
13458,Masculino,NaN,Colombia,Envigado,San José,0,NaN,NaN,NaN,NaN,...,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN,0.0
13459,Femenino,Envigado,Colombia,Envigado,La Mina,3,Soltero(a),21) EPS Sura,N0,Urbana,...,Activo,NaT,MEDIA TÉCNICA,MEDIA TÉCNICA TÉCNICO EN MARKETING Y VENTA DE ...,2026-MT,ONCE,16,NaN,NaN,0.0


### Regla de continuidad para los estados ambiguos
Para `Inactivo`, `Aplazamiento Semestre` y `Aplazamiento Actividades` aplicamos la regla de
los 2 años: si la última matrícula fue hace **más de 2 años** se considera deserción (`1`);
si fue hace **menos de 2 años**, no deserción (`0`). Si no hay fecha válida, se excluye.

In [37]:
FECHA_CORTE = pd.Timestamp('2022-06-29').normalize()
LIMITE_2A   = FECHA_CORTE - pd.DateOffset(years=2)

estados_temporales = ['Inactivo', 'Aplazamiento Semestre', 'Aplazamiento Actividades']
mask_temporal = df['estado'].isin(estados_temporales)

# Más de 2 años sin volver a matricularse -> deserción (1)
df.loc[mask_temporal & (df['fecha_de_matricula'] <  LIMITE_2A), 'desercion'] = 1
# Menos de 2 años -> aún podría reingresar -> no deserción (0)
df.loc[mask_temporal & (df['fecha_de_matricula'] >= LIMITE_2A), 'desercion'] = 0
# Sin fecha válida (NaT) en estado temporal -> se queda NaN (no se puede aplicar la regla)

# --- Diagnóstico: cómo se repartieron los estados temporales ---
print('Fecha de corte:', FECHA_CORTE.date(), '| Umbral 2 años:', LIMITE_2A.date())
print()
for est in estados_temporales:
    sub = df[df['estado'] == est]
    a_1 = (sub['fecha_de_matricula'] <  LIMITE_2A).sum()
    a_0 = (sub['fecha_de_matricula'] >= LIMITE_2A).sum()
    nat = sub['fecha_de_matricula'].isna().sum()
    print(f'{est:<26} +2años->1: {a_1:>5} | <2años->0: {a_0:>4} | sin fecha: {nat}')

# Verificar que no quedó ningún estado sin clasificar por error de tipeo
clasificados = set(mapa_desercion) | set(estados_temporales)
sin_clasificar = set(df['estado'].dropna().unique()) - clasificados
print('\nEstados sin clasificar (debe ser vacío):', sin_clasificar)

Fecha de corte: 2022-06-29 | Umbral 2 años: 2020-06-29

Inactivo                   +2años->1:  3000 | <2años->0: 1715 | sin fecha: 123
Aplazamiento Semestre      +2años->1:   171 | <2años->0:  114 | sin fecha: 0
Aplazamiento Actividades   +2años->1:     3 | <2años->0:    5 | sin fecha: 0

Estados sin clasificar (debe ser vacío): set()


### Resumen del target resultante
Definición elegida: cuántas filas quedan etiquetadas,
cuántas se excluyen y el balance de clases (relevante para el modelado posterior).

In [52]:
n_total = len(df)
n_etiquetadas = df['desercion'].notna().sum()
n_excluidas = df['desercion'].isna().sum()

print(f"Filas totales:      {n_total}")
print(f"Etiquetadas:        {n_etiquetadas}")
print(f"Excluidas (NaN):    {n_excluidas}")
print()
print("Distribución del target (solo etiquetadas):")
print(df['desercion'].value_counts(dropna=False))
print()
tasa = df['desercion'].mean()
print(f"Tasa de deserción:  {tasa:.1%}")
print()

Filas totales:      13397
Etiquetadas:        13036
Excluidas (NaN):    361

Distribución del target (solo etiquetadas):
desercion
0.0    8451
1.0    4585
NaN     361
Name: count, dtype: int64

Tasa de deserción:  35.2%



## Revisión de duplicados

El dataset son *matrículas*, no estudiantes únicos. Se revisará filas
completamente duplicadas.

In [53]:
n_dup = df.duplicated().sum()
print("Filas duplicadas exactas:", n_dup)
df = df.drop_duplicates().reset_index(drop=True)
print("Shape tras eliminar duplicados:", df.shape)

Filas duplicadas exactas: 0
Shape tras eliminar duplicados: (13397, 24)


Anteriormente se detectaron 64 filas duplicadas, pero al ejecutar el código "df = df.drop_duplicates().reset_index(drop=True)" se eliminan las filas duplicadas.

In [54]:
filas_duplicadas = df[df.duplicated(keep=False)].sort_values(by=df.columns.tolist())
print(f"Total de filas duplicadas: {len(filas_duplicadas)}")
print()
filas_duplicadas

Total de filas duplicadas: 0



,genero,municipio_de_nacimiento,pais,municipio_direccion,barrio_direccion,estrato,estado_civil,eps,grupo_sisben,zona,...,fecha_de_matricula,jornada,programa,periodo,nivel,edad,anio_matricula,mes_matricula,desercion,estrato_imputado


## 7. Columnas con exceso de nulos

> **Retroalimentación:** algunas variables se caerán por su alto % de nulos; el dataset se
> reducirá en columnas. Aquí identificamos y descartamos las que están **casi vacías**
> (la decisión final sobre el resto se toma en el EDA con pruebas estadísticas).

In [57]:
nulos_pct = (df.isnull().mean() * 100).round(1).sort_values(ascending=False)
print("% de nulos por columna:")
print(nulos_pct.head(8))

# Descartar columnas con >90% de nulos (aportan casi nula información)
umbral = 70
a_descartar = nulos_pct[nulos_pct > umbral].index.tolist()
print(f"\nColumnas descartadas (>{umbral}% nulos): {a_descartar}")
df = df.drop(columns=a_descartar)
print("Shape tras descartar:", df.shape)

% de nulos por columna:
nivel_de_formacion     24.9
medio_de_transporte    24.7
eps                    22.6
ocupacion              14.2
estado_civil           10.7
barrio_direccion        9.8
zona                    9.5
anio_matricula          9.0
dtype: float64

Columnas descartadas (>70% nulos): []
Shape tras descartar: (13397, 23)


## Guardar los datos limpios

Guardar el dataset limpio, estandarizado y con la variable objetivo.

In [58]:
from getpass import getpass

ruta_salida = processed_dir + 'matriculas_limpias.csv'
df.to_csv(ruta_salida, index=False)
print("Guardado en:", ruta_salida)
print("Dimensiones finales:", df.shape)

!git config --global user.email "betancourt1659@gmail.com"
!git config --global user.name  "LuisBetancourt11"

usuario = "LuisBetancourt11"
repo    = "Proyecto_Ciencia_Datos"
token   = getpass("Access Token: ")

!cd {repo} && git add data/processed/matriculas_limpias.csv && \
  git commit -m "Limpieza corregida)" && \
  git remote set-url origin https://{usuario}:{token}@github.com/{usuario}/{repo}.git && \
  git push origin main

Guardado en: Proyecto_Ciencia_Datos/data/processed/matriculas_limpias.csv
Dimensiones finales: (13397, 23)
Access Token: ··········
[main 716a81a] Limpieza corregida)
 1 file changed, 13398 insertions(+), 13398 deletions(-)
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 2 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 269.07 KiB | 2.08 MiB/s, done.
Total 5 (delta 2), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
To https://github.com/LuisBetancourt11/Proyecto_Ciencia_Datos.git
   95a463f..716a81a  main -> main
